From c968a56a7b8b5afd01889ea63272677a90945078 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 052/296] Reset surface_texture field to avoid crash in JNI
 MediaPlayer

When invoke setVideoSurface in JNI MediaPlayer, it will decrease
reference of old surface texture, and then set new surface_texture
field after get new surface. But new surface_texture field will never
set if surface already being released or surface texture binding
failed. When next time setVideoSurface invoked and try to get video
surface texture, it will cause crash for accessing illegal address.

Reset surface_texture field when surface fails to bind surface
texture or surface has already been released.
---
 media/jni/android_media_MediaPlayer.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/media/jni/android_media_MediaPlayer.cpp b/media/jni/android_media_MediaPlayer.cpp
index c8ec7c4a3ba..48874a55afb 100644
--- a/media/jni/android_media_MediaPlayer.cpp
+++ b/media/jni/android_media_MediaPlayer.cpp
@@ -341,12 +341,14 @@ setVideoSurface(JNIEnv *env, jobject thiz, jobject jsurface, jboolean mediaPlaye
         if (surface != NULL) {
             new_st = surface->getIGraphicBufferProducer();
             if (new_st == NULL) {
+                env->SetLongField(thiz, fields.surface_texture, 0);
                 jniThrowException(env, "java/lang/IllegalArgumentException",
                     "The surface does not have a binding SurfaceTexture!");
                 return;
             }
             new_st->incStrong((void*)decVideoSurfaceRef);
         } else {
+            env->SetLongField(thiz, fields.surface_texture, 0);
             jniThrowException(env, "java/lang/IllegalArgumentException",
                     "The surface has been released");
             return;
-- 
2.11.0

